src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,571 @@
+        os.path.join(collection_source_dir, d)
+        for d in os.listdir(collection_source_dir)
+        if os.path.isdir(os.path.join(collection_source_dir, d))
+    ]
+
+    for assembly_dir in assembly_dirs:
+
+        assembly_file_list = [
+            f
+            for f in os.listdir(assembly_dir)
+            if os.path.isfile(os.path.join(assembly_dir, f))
+            and os.path.join(assembly_dir, f).endswith(tuple(upload_file_ext))
+        ]
+
+        if len(assembly_file_list) != 1:
+            raise ValueError(f"One and only one assembly file that ends with {upload_file_ext} "
+                    f"must be present in {assembly_dir} directory")
+
+        count += 1
+        assembly_name = assembly_file_list[0]
+
+        _, uploaded = _read_upload_status_yaml_file(upload_env_key, workspace_id, assembly_dir, assembly_name)
+
+        if uploaded:
+            print(
+                f"Assembly {assembly_name} already exists in workspace {workspace_id}. Skipping."
+            )
+            continue
+
+        wait_to_upload_assemblies[assembly_name] = assembly_dir
+
+    return count, wait_to_upload_assemblies
+
+
+def _prepare_skd_job_dir_to_upload(job_dir: str, wait_to_upload_assemblies: dict[str, str]) -> str:
+    """
+    Prepare SDK job directory to upload.
+    """
+    data_dir = _sanitize_data_dir(job_dir)
+    for assembly_file, assembly_dir in wait_to_upload_assemblies.items():
+        src_file = _get_source_file(assembly_dir, assembly_file)
+        dest_file = os.path.join(data_dir, assembly_file)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+
+    return data_dir
+
+
+def _create_entries_in_sourcedata_workspace(
+        upload_env_key: str,
+        workspace_id: int,
+        assembly_names: list[str],
+        assembly_name_to_dir: dict[str, str],
+        assembly_name_to_upa: dict[str, str],
+        output_dir: str,
+) -> list[str]:
+    """
+    Create a standard entry in sourcedata/workspace for each assembly.
+    Hardlink to the original assembly file in sourcedata to avoid duplicating the file.
+    Update the uploaded.yaml file in the genome directory with assembly names and upa info.
+    """
+    upas = list()
+    for assembly_name in assembly_names:
+        try:
+            assembly_dir = assembly_name_to_dir[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name}") from e
+
+        src_file = _get_source_file(assembly_dir, assembly_name)
+
+        try:
+            upa = assembly_name_to_upa[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name} from target workspace") from e
+        
+        upa_dir = os.path.join(output_dir, upa)
+        os.makedirs(upa_dir, exist_ok=True)
+        upas.append(upa)
+
+        dest_file = os.path.join(upa_dir, assembly_name)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
```

Comment:
Updated!

Reply:
That's the same way the workspace downloader names them?

Reply:
Sorry for the confusion. I forget to put ``, and that's why `<upa>` is missing from the expression above.

In the workspace downloader, we have `<upa>.fa`
In the current workspace uploader, we have `<upa>.fna.gz` file

They have the same naming convention, but slightly different file extensions.

I assume .fa and .fna shouldn't affect the final result since they have the same content. Please correct me if I am wrong @MrCreosote @Tianhao-Gu

Reply:
I don't know if that would affect the tool runners and tools or not. Is there a reason not to just use the same extension?

Reply:
This seems like another case where maybe the uploader and downloader should share a little code so they're consistent

Reply:
>  Is there a reason not to just use the same extension?

**Uploader**: GTDB files that we downloaded from NCBI FTP server have .fna.gz extensions in nature. And we are now trying to upload these files to the target workspace.

**Downloader**: We directly downloaded fasta files from the target workspace.

I personally think is more efficient to upload .fna.gz file and let the tool runners handle unzipping. Please let me know if you think we should take a different approach.

Reply:
let me ask the question a different way - will the subsequent steps in the pipeline work given the different filenames / formats?

Reply:
IMO, set unzip=True in the tool runners should do the job.

Reply:
So do changes need to be made to the tool runners?

Reply:
I think ppl run GTDB files need to manually set unzip=True based on my understanding. 
Can you confirm that @Tianhao-Gu? 

If that's the case, no changes need to made to the tool runners in this PR. But ideally, I would prefer something more general.

Reply:
it was not set by the person who run the tool, rather it was set by how the tool_runner was called. 

For example, the mash tool doesn't handle .gz files, so we want to unzip files before executing, and delete the unzipped files after execution.
https://github.com/kbase/collections/blob/main/src/loaders/compute_tools/mash/mash.py#L45

But GTDB_TK the tool itself knows how to handle .gz files, so we do not unzip it. 
https://github.com/kbase/collections/blob/main/src/loaders/compute_tools/checkm2/checkm2.py#L59

Reply:
Thank you! @Tianhao-Gu

No changes need to be made to the tool runners then. @MrCreosote